src/csharp/Grpc.Examples/MathGrpc.cs

```diff
@@ -27,10 +27,33 @@ public static partial class Math
   {
     static readonly string __ServiceName = "math.Math";
 
-    static readonly grpc::Marshaller<global::Math.DivArgs> __Marshaller_math_DivArgs = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.DivArgs.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.DivReply> __Marshaller_math_DivReply = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.DivReply.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.FibArgs> __Marshaller_math_FibArgs = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.FibArgs.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.Num> __Marshaller_math_Num = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.Num.Parser.ParseFrom);
+    static void __Helper_SerializeMessage(global::Google.Protobuf.IMessage message, grpc::SerializationContext context)
+    {
+      #if !GRPC_DISABLE_PROTOBUF_BUFFER_SERIALIZATION
+      if (message is global::Google.Protobuf.IBufferMessage)
+      {
+        context.SetPayloadLength(message.CalculateSize());
+        global::Google.Protobuf.MessageExtensions.WriteTo(message, context.GetBufferWriter());
+        context.Complete();
+        return;
+      }
+      #endif
+      context.Complete(global::Google.Protobuf.MessageExtensions.ToByteArray(message));
+    }
+
+    static T __Helper_DeserializeMessage<T>(grpc::DeserializationContext context, global::Google.Protobuf.MessageParser<T> parser) where T : global::Google.Protobuf.IMessage<T>
+    {
+      #if !GRPC_DISABLE_PROTOBUF_BUFFER_SERIALIZATION
+      return parser.ParseFrom(context.PayloadAsReadOnlySequence());
```

Comment:
One danger I've noticed:
While in the __Helper_SerializeMessage method we can quickly check whether the message is IBufferMessage (which allows us to fall back to the old way of serializing in cases where the user still has the old generated code),  in case of __Helper_DeserializeMessage we don't have this check, so if user hasn't regenerated the protobuf code, deserialization will fail (old generated code cannot parse from a ParseContext if there's no codedInputStream instance to fallback to when parsing).

Btw, checking "message is IBufferWriter" is not bulletproof (because some of the nested messages still might not support buffer serialization), but it's a pretty good indication of whether it's safe to use the buffer serialization or not.

Reply:
Basically what we want is something in the sense of `typeof(global::Google.Protobuf.IBufferMessage).IsAssignableFrom(typeof(T))` but without needing to use reflection (which I assume would be slow).

Reply:
I think you could cache the result of `typeof(global::Google.Protobuf.IBufferMessage).IsAssignableFrom(typeof(T))` in a generic static type.

```cs
private static class BufferMessageHelper<T>
{
    public static bool IsBufferMessage = typeof(global::Google.Protobuf.IBufferMessage).IsAssignableFrom(typeof(T));
}

static T __Helper_DeserializeMessage<T>(grpc::DeserializationContext context, global::Google.Protobuf.MessageParser<T> parser) where T : global::Google.Protobuf.IMessage<T>
{
    if (BufferMessageHelper<T>.IsBufferMessage)
    {
        // ...
    }
}
```

Reply:
That would work. I used that snippet with some modifications.